keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,211 @@
+---
+kep-number: 35
+title: Inline CSI Volumes Support
+authors:
+  - "@vladimirvivien"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-aaa
+  - sig-bbb
+reviewers:
+  - "@thockin"
+  - "@saad-ali"
+  - "@msau42"
+  - "@jsafrane"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-01-22
+last-updated: 2019-01-22
+status: provisional
+---
+
+# Inline CSI volumes support
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [User Stories](#user-stories)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+
+
+## Summary
+Currently, volumes that are backed by CSI drivers can only be used with the `PersistentVolume` and `PersistentVolumeClaim` objects. This proposal is to implement support for the ability to nest CSI volume declarations within pod specs, as is supported alrady by existing in-tree storage providers.
+
+This KEP started life as [feature #2273](https://github.com/kubernetes/community/pull/2273).  Please follow that link for historical context.
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identy, variables or similar information inside pods. 
```

Comment:
This feature is much needed for this: https://github.com/deislabs/secrets-store-csi-driver

Reply:
And https://github.com/kubernetes-csi/drivers/issues/85